src/app/reporting/ReportSchedulerImpl.cpp

```diff
@@ -0,0 +1,206 @@
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app/InteractionModelEngine.h>
+#include <app/reporting/ReportSchedulerImpl.h>
+
+namespace chip {
+namespace app {
+namespace reporting {
+
+using Seconds16       = System::Clock::Seconds16;
+using Milliseconds32  = System::Clock::Milliseconds32;
+using Timeout         = System::Clock::Timeout;
+using Timestamp       = System::Clock::Timestamp;
+using ReadHandlerNode = ReportScheduler::ReadHandlerNode;
+
+/// @brief Callback called when the report timer expires to schedule an engine run regardless of the state of the ReadHandlers, as
+/// the engine already verifies that read handlers are reportable before sending a report
+static void ReportTimerCallback()
+{
+    InteractionModelEngine::GetInstance()->GetReportingEngine().ScheduleRun();
+}
+
+ReportSchedulerImpl::ReportSchedulerImpl(TimerDelegate * aTimerDelegate) : ReportScheduler(aTimerDelegate)
+{
+    VerifyOrDie(nullptr != mTimerDelegate);
+}
+
+/// @brief When a ReadHandler is added, register it, which will schedule an engine run on the max interval
+void ReportSchedulerImpl::OnReadHandlerAdded(ReadHandler * aReadHandler)
+{
+    RegisterReadHandler(aReadHandler);
+}
+
+/// @brief When a ReadHandler becomes reportable, schedule, verifies if the handler is the min interval is elapsed. If not,
+/// reschedule the report to happen when the min interval is elapsed. If it is, schedule an engine run.
+void ReportSchedulerImpl::OnBecameReportable(ReadHandler * aReadHandler)
+{
+    ReadHandlerNode * node = FindReadHandlerNode(aReadHandler);
+    VerifyOrReturn(nullptr != node);
+    if (node->IsReportableNow())
+    {
+        // If the handler is reportable now, just schedule a report immediately
+        InteractionModelEngine::GetInstance()->GetReportingEngine().ScheduleRun();
+    }
+    else
+    {
+        Milliseconds32 newTimeout =
+            Milliseconds32(node->GetMinInterval().count() - System::SystemClock().GetMonotonicTimestamp().count());
+        // If the handler is not reportable now, schedule a report for the min interval
+        ScheduleReport(newTimeout, aReadHandler);
+    }
+}
+
+void ReportSchedulerImpl::OnReportSent(ReadHandler * apReadHandler)
+{
+    ReadHandlerNode * node = FindReadHandlerNode(apReadHandler);
+    VerifyOrReturn(nullptr != node);
+    // Schedule callback for max interval
+    node->SetIntervalsTimeStamp(apReadHandler);
+    Milliseconds32 newTimeout =
+        Milliseconds32(node->GetMaxInterval().count() - System::SystemClock().GetMonotonicTimestamp().count());
```

Comment:
Is this the right computation of timeout? Interval is a duration. ScheduleReport documents that it takes a timeout (e.g. a number of seconds duration, and not a time point). Why are we subtracting the GetMonotonicTimestamp here? This seems incorrect.

Reply:
The "Interval" naming was confusing. `newTimeout` is obtained by subtracting the Timestamp at which the max interval expires to the current timestamp.